### ISSUE TITLE
Ignoring `this` checking as parameter for instance method hook for `Hook`

### ISSUE THREAD

**Issue by SvetlovA** [AngelDE98/MonoMod#119]
### Description
Do you plan to move classical `Detour` to `reorganize` branch?
If not, maybe it's a good idea to enable/disable `this` ignoring for source and target in `Hook`. I frequently have situation where I need to ignore `this` in both. And without this feature I always have an error "Target method is not compatible with source method" when try to hook instance method. But if I add `this` to declaring hook I have a problem that I need to call this hook without explicitly put `this`.

### Example
```cs
public sealed class Hook : IDetour, IDisposable {
....
 private MethodInfo PrepareRealTarget(object? target, out MethodInfo trampoline, out DataScope<DynamicReferenceCell> scope)
 {
      var srcSig = MethodSignature.ForMethod(Source, ignoreThis:  Config.SourceIgnoreThis);
      var dstSig = MethodSignature.ForMethod(Target, ignoreThis: Config.TargetIgnoreThis); 
....
```

**Comment by nike4613** [comment]
`Detour` actually used to be in `reorganize`. It was removed because its functionality was a strict subset of `Hook`'s. 

In terms of ignoring this, I am fairly certain you don't actually want to do that. The checks that exist are there to ensure that the hook can actually be safely applied; if it rejects something, its likely because it would not be reliable.

Can you share more about your specific usecase?

**Comment by SvetlovA** [comment]
Here is my small example of usecase:

```cs
public class SumInstance
{
    public int Sum(int param)
    {
        return param + param;
    }
}

public class MultiplyInstance
{
    public int Multiply(int param)
    {
        return param * param;
    }
}

public class Tests
{
    [Test]
    public void TestHook()
    {
        const int testParam = 3;
        const int expectedSum = 6;
        const int expectedMultiply = 9;

        var sumMethodInfo = typeof(SumInstance).GetMethod(nameof(SumInstance.Sum));
        var multiplyMethodInfo = typeof(MultiplyInstance).GetMethod(nameof(MultiplyInstance.Multiply));

        var sumInstance = new SumInstance();

        using (var hook = new Hook(sumMethodInfo, multiplyMethodInfo, new MultiplyInstance()))
        {
            Assert.AreEqual(expectedMultiply, sumInstance.Sum(testParam));
        }

        Assert.AreEqual(expectedSum, sumInstance.Sum(testParam));
    }
}
```
Error:
```
System.ArgumentException : Target method is not compatible with source method
   at MonoMod.RuntimeDetour.Hook.PrepareRealTarget(Object target, MethodInfo& trampoline, DataScope`1& scope)
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase source, MethodInfo target, Object targetObject, IDetourFactory factory, DetourConfig config, Boolean applyByDefault)
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase source, MethodInfo target, Object targetObj, DetourConfig config, Boolean applyByDefault)
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase source, MethodInfo target, Object targetObj, DetourConfig config)
   at MonoMod.RuntimeDetour.Hook..ctor(MethodBase source, MethodInfo target, Object targetObj)
   at MonoModHookTest.Tests.TestHook() in D:\MyProjects\MonoModHookTest\MonoModHookTest\Tests.cs:line 35
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```
The same behavior is for current main version from `master`, but for simple `Detour` without validation all works as expected

**Comment by nike4613** [comment]
That is the correct behavior. `Hook` does not, and has never, supported dropping parameters. The `this` parameter is a parameter. We don't support that because that would either enable silently unsafe-casting one type to another, or bizarre and unintuitive behavior. For this example. the supported workaround to to add a first parameter to `MultiplyInstance.Multiply` with type `SumInstance`. Alternately, `MultiplyInstance.Multiply` can be made static, and everything continues to apply. If you do not control `MultiplyInstance`, create or generate a wrapper. If you need the unsafe-casting, use `Unsafe.As` yourself.

---

Old RuntimeDetour did not do enough checking, especially around `Detour. It would silently do things which could easily cause either GC holes or more direct segfaults. The validation is correct.

**Comment by nike4613** [comment]
All that being said, it *may* be worth implementing support for your usecase. If that were to be done, I'd want it to be implicit, based on the method parameters like `orig` delegates, rather than specified via config. Unfortunately, this would be a non-tivial amount of work, and would (in the general case) introduce a level of overhead which I do not believe that MonoMod should have.

Specifically, in order to correctly invoke the `orig` method, the `Hook` wrapper must construct a new delegate which invokes the trampoline with the `this` param that it was given, which implies an allocation. This same effect would then be present throughout a detour chain, introducing many allocations where there previously were none.
 
---

In summary, I think that the best solution to this is for you to create your own helpers to explicitly drop the first argument. It might be worth exposing them in MonoMod itself somehow, but 1) putting it in `Hook` directly would introduce overhead which may not be obvious and may be unwanted or unneeded, and 2) placing it outside `Hook` feels inconsistent, given that other features which appear similar are built in to `Hook`.

**Comment by SvetlovA** [comment]
Thank you very much for helping with understanding.
Option with adding a first parameter to `MultiplyInstance.Multiply` with type `SumInstance` is fine for me.
